persistent/test/Database/Persist/TH/EntityHaddockSpec.hs

```diff
@@ -0,0 +1,20 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Database.Persist.TH.EntityHaddockSpec (spec) where
+
+import TemplateTestImports
+
+#if MIN_VERSION_template_haskell(2,18,0)
+import Database.Persist.TH.CommentSpec (CommentModel (..))
+import Language.Haskell.TH (DocLoc (DeclDoc), getDoc, runQ)
+
+spec :: Spec
+spec = describe "EntityHaddockSpec" $ do
+    it "generates entity Haddock" $ do
+        maybeDoc <- runQ $ getDoc (DeclDoc 'CommentModel)
```

Comment:
Grepping `ghc` codebase for `Template Haskell failure` gets us this:

```haskell
badIO :: String -> IO a
badIO op = do   { qReport True ("Can't do `" ++ op ++ "' in the IO monad")
                ; fail "Template Haskell failure" }
```

Which is kind of an awful error - the actual reason is interleaved within the [test output, here](https://github.com/blujupiter32/persistent/actions/runs/5395910491/jobs/9798925953?pr=5#step:16:249).

I think what you'll want to do here is splice the doc into the file, and then you can write assertions on the doc in a spec. Like,

```haskell
[d|
commentModelDoc :: String
commentModelDoc = 
  $(lift =<< getDoc (DeclDoc 'CommentModel))
|]

spec = describe "EntityHaddockSpec" $ do
    it "generates entity haddock" $ do
      commentModelDoc `shouldBe` "..."
```

Or, if you want a *compilation* failure, you could write the whole "test" in the splice.

```haskell
module Test where
 
import Doc

do
    maybeDoc <- getDoc (DeclDoc 'CommentModel)
    case maybeDoc of
        Nothing -> reportError "Expected doc for CommentModel"
        Just doc -> do
            let expected = "..."
            when (doc /= expected) $ do
                reportError "Expected doc to be different..."

    pure [] -- to satisfy the `Q [Dec]` type for a top level splice   
```

But this is probably much less easy to write, and won't get you nice error message rendering. But it *will* fail faster, so idk. `hspec-compilation` is begging to be born I guess lol.

Reply:
I've completed `EntityHaddockSpec` using the former method. However, the [test output](https://github.com/blujupiter32/persistent/actions/runs/5403252476/jobs/9815790608?pr=5#step:16:453) shows that it only finds the field comment added using `putDoc` - attempting to get the Haddock for `CommentModel` (added using `withDecDoc`) returns `Nothing`.

Reply:
Now fixed - I'd put `'CommentModel` (constructor `Name`) instead of `''CommentModel` (type `Name`) in the splice.

Reply:
Oh goodness - I guess you *can* write, like,

```haskell
-- | commentary for type
data Thing = 
    -- | commentary for constructor
    MkThing 
```

Great catch 😄